pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/MissingConsumingSegmentFinder.java

```diff
@@ -0,0 +1,151 @@
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.helix.AccessOption;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.model.IdealState;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.apache.pinot.common.metadata.ZKMetadataProvider;
+import org.apache.pinot.common.metrics.ControllerGauge;
+import org.apache.pinot.common.metrics.ControllerMeter;
+import org.apache.pinot.common.metrics.ControllerMetrics;
+import org.apache.pinot.common.utils.LLCSegmentName;
+import org.apache.pinot.spi.utils.CommonConstants.Helix.StateModel.SegmentStateModel;
+import org.apache.zookeeper.data.Stat;
+
+
+/**
+ * For a given table, this class finds out if there is any partition group for which there's no consuming segment in
+ * ideal state. If so, it emits two metrics:
+ *   - Number of missing consuming segments
+ *   - Maximum duration (in minutes) that a partition hasn't had a consuming segment
+ */
+public class MissingConsumingSegmentFinder {
+
+  private String _realtimeTableName;
+  private ControllerMetrics _controllerMetrics;
+  private SegmentMetadataFetcher _segmentMetadataFetcher;
+
+  public MissingConsumingSegmentFinder(String realtimeTableName, ZkHelixPropertyStore<ZNRecord> propertyStore,
+      ControllerMetrics controllerMetrics) {
+    _realtimeTableName = realtimeTableName;
+    _controllerMetrics = controllerMetrics;
+    _segmentMetadataFetcher = new SegmentMetadataFetcher(propertyStore, controllerMetrics);
+  }
+
+  @VisibleForTesting
+  MissingConsumingSegmentFinder(String realtimeTableName, SegmentMetadataFetcher segmentMetadataFetcher) {
+    _realtimeTableName = realtimeTableName;
+    _segmentMetadataFetcher = segmentMetadataFetcher;
+  }
+
+  public void findAndEmitMetrics(IdealState idealState) {
+    MissingSegmentInfo info = findMissingSegments(idealState.getRecord().getMapFields(), Instant.now());
+    _controllerMetrics.setValueOfTableGauge(_realtimeTableName, ControllerGauge.MISSING_CONSUMING_SEGMENT_COUNT,
+        info._count);
+    _controllerMetrics
+        .setValueOfTableGauge(_realtimeTableName, ControllerGauge.MISSING_CONSUMING_SEGMENT_MAX_DURATION_MINUTES,
+            info._maxDurationInMinutes);
+  }
+
+  @VisibleForTesting
+  MissingSegmentInfo findMissingSegments(Map<String, Map<String, String>> idealStateMap, Instant now) {
+    // create the maps
+    Map<Integer, LLCSegmentName> partitionGroupIdToLatestConsumingSegmentMap = new HashMap<>();
+    Map<Integer, LLCSegmentName> partitionGroupIdToLatestCompletedSegmentMap = new HashMap<>();
+    idealStateMap.forEach((segmentName, instanceToStatusMap) -> {
+      LLCSegmentName llcSegmentName = LLCSegmentName.of(segmentName);
+      if (llcSegmentName != null) { // Skip the uploaded realtime segments that don't conform to llc naming
+        if (instanceToStatusMap.containsValue(SegmentStateModel.CONSUMING)) {
+          updateMap(partitionGroupIdToLatestConsumingSegmentMap, llcSegmentName);
```

Comment:
have you on purpose omitted the cases where there's no CONSUMING segment, but the ValidationManager will be fixing it?

Reply:
Like 1) OFFLINE state segment, or 2) when it was the very first segment that was deleted, hence no completed segment recorded 3) new partition
I feel this should be a generic metric which says how many are not CONSUMING, and be oblivious to what is correctable ad what's not

Reply:
1) My understanding is that if a consuming segment encounters a permanent error, it calls controller to mark it as OFFLINE in ideal state. So for these cases, the last consumed time is when the previous completed segment went to ONLINE state. 
2) We can ignore the new partition cases as Segment Validation Manager is going to create a new consuming segment for them. If we wanted to detect them, then we'd need to fetch the metadata for the stream which may be problematic (please refer to next comment).

Reply:
And for the new partition case, my strong preference would still be to include them here so this becomes a reliable metric covering all cases. It would be hard to reason about and explain to users about the cases this includes and excludes (like if they delete a CONSUMING segment that had seqId 3, we would detect it, but if they deleted CONSUMING segment with segId 0, we wouldnt detect it, and then proceed to explain how that is okay to not detect because validation manager would anyway fix it.)
But if keeping this lightweight and free of stream connections and stream error handling is a string requirement, I guess it's fine.